plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftConnector.java

```diff
@@ -524,6 +534,42 @@ private void idleCheServer(ServerIdleEvent event) {
     return portsToRefName;
   }
 
+  private Subject subjectForContainerId(String containerId) {
+    String workspaceId = containerIdToWorkspaceId.get(containerId);
+    if (workspaceId != null) {
+      Subject subject = workspaceSubjectRegistry.getWorkspaceStarter(workspaceId);
```

Comment:
Why we need to get Subject from cache instead of ```EnvironmentContext.getCurrent().getSubject();```

Reply:
to manage the cases when `OpenshiftConnector` has to to some action on a workspace, triggered by a batch, non-UI event:
1. start/stop of the che-server (which will stop ll workspaces during the shutdown procedure, which in turn will call `OpenshiftConnector.removeImage()` and `OpenshiftConnector.removeContainer()`.
2. stop of the current workspace by the `activity-checker` due to workspace idling, which will also call `OpenshiftConnector.removeImage()` and `OpenshiftConnector.removeContainer()`.

In cases 1 and 2, the `EnvironmentContext.getCurrent().getSubject()` doesn't contain a valid subject that can be used to connect the Openshift cluster/namespace hosting the workspace.

3. Start of the Bayesian language server: in this case, calling `EnvironmentContext.getCurrent().getSubject()` on the wsmaster during a request coming from the wsagent would return the *machine token* instead of the Keycloak token in `subject.getToken()`, which is also wrong.

Reply:
it would be great if subject is not cached
why batch jobs can't send the required data ? or why associated subject is not propagated ? on removeContainer ?

Reply:
@benoitf in those 3 scenarios, currently (before this PR) the required user information (contained in the subject) is available nowhere in the wsmaster application. And since these 3 scenarios are triggered by some non-interactive event (idling, tomcat shutdown hook, internal event indirectly triggered by the language server registry), we have no way to pass or propagate the user information.

To be clear workspace stop initiated by the user itself from the Javascript UIs (GWT or Dashboard) was working great.